// @typegpu/color/index.mjs
import tgpu from 'typegpu';
import { vec3f, f32, mat3x3f, struct } from 'typegpu/data';
import { floor, max, min, mul, select, sub, pow, gt, add, length, abs, sign, sqrt, clamp, mix } from 'typegpu/std';

const hsvToRgb = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((hsv) => {
  const h = hsv.x;
  const s = hsv.y;
  const v = hsv.z;
  const i = floor(h * 6);
  const f = h * 6 - i;
  const p = v * (1 - s);
  const q = v * (1 - f * s);
  const t = v * (1 - (1 - f) * s);
  let r = f32(0);
  let g = f32(0);
  let b = f32(0);
  if (i % 6 === 0) {
    r = v;
    g = t;
    b = p;
  } else if (i % 6 === 1) {
    r = q;
    g = v;
    b = p;
  } else if (i % 6 === 2) {
    r = p;
    g = v;
    b = t;
  } else if (i % 6 === 3) {
    r = p;
    g = q;
    b = v;
  } else if (i % 6 === 4) {
    r = t;
    g = p;
    b = v;
  } else {
    r = v;
    g = p;
    b = q;
  }
  return vec3f(r, g, b);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"hsv"}],"body":[0,[[13,"h",[7,"hsv","x"]],[13,"s",[7,"hsv","y"]],[13,"v",[7,"hsv","z"]],[13,"i",[6,"floor",[[1,"h","*",[5,"6"]]]]],[13,"f",[1,[1,"h","*",[5,"6"]],"-","i"]],[13,"p",[1,"v","*",[1,[5,"1"],"-","s"]]],[13,"q",[1,"v","*",[1,[5,"1"],"-",[1,"f","*","s"]]]],[13,"t",[1,"v","*",[1,[5,"1"],"-",[1,[1,[5,"1"],"-","f"],"*","s"]]]],[12,"r",[6,"f32",[[5,"0"]]]],[12,"g",[6,"f32",[[5,"0"]]]],[12,"b",[6,"f32",[[5,"0"]]]],[11,[1,[1,"i","%",[5,"6"]],"===",[5,"0"]],[0,[[2,"r","=","v"],[2,"g","=","t"],[2,"b","=","p"]]],[11,[1,[1,"i","%",[5,"6"]],"===",[5,"1"]],[0,[[2,"r","=","q"],[2,"g","=","v"],[2,"b","=","p"]]],[11,[1,[1,"i","%",[5,"6"]],"===",[5,"2"]],[0,[[2,"r","=","p"],[2,"g","=","v"],[2,"b","=","t"]]],[11,[1,[1,"i","%",[5,"6"]],"===",[5,"3"]],[0,[[2,"r","=","p"],[2,"g","=","q"],[2,"b","=","v"]]],[11,[1,[1,"i","%",[5,"6"]],"===",[5,"4"]],[0,[[2,"r","=","t"],[2,"g","=","p"],[2,"b","=","v"]]],[0,[[2,"r","=","v"],[2,"g","=","p"],[2,"b","=","q"]]]]]]]],[10,[6,"vec3f",["r","g","b"]]]]],"externalNames":["floor","f32","vec3f"]},
              externals: () => ({floor, f32, vec3f}),
            }) && $.f)({}))), "hsvToRgb"));
const rgbToHsv = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((rgb) => {
  const r = rgb.x;
  const g = rgb.y;
  const b = rgb.z;
  const maxC = max(r, max(g, b));
  const minC = min(r, min(g, b));
  const delta = f32(maxC - minC);
  let h = f32(0);
  let s = f32(0);
  if (maxC === 0) {
    s = 0;
  } else {
    s = delta / maxC;
  }
  const v = maxC;
  if (maxC === minC) {
    h = 0;
  } else if (maxC === r) {
    let cond = f32(0);
    if (g < b) {
      cond = 6;
    } else {
      cond = 0;
    }
    h = g - b + delta * cond;
    h /= 6 * delta;
  } else if (maxC === g) {
    h = b - r + delta * 2;
    h /= 6 * delta;
  } else if (maxC === b) {
    h = r - g + delta * 4;
    h /= 6 * delta;
  }
  return vec3f(h, s, v);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"rgb"}],"body":[0,[[13,"r",[7,"rgb","x"]],[13,"g",[7,"rgb","y"]],[13,"b",[7,"rgb","z"]],[13,"maxC",[6,"max",["r",[6,"max",["g","b"]]]]],[13,"minC",[6,"min",["r",[6,"min",["g","b"]]]]],[13,"delta",[6,"f32",[[1,"maxC","-","minC"]]]],[12,"h",[6,"f32",[[5,"0"]]]],[12,"s",[6,"f32",[[5,"0"]]]],[11,[1,"maxC","===",[5,"0"]],[0,[[2,"s","=",[5,"0"]]]],[0,[[2,"s","=",[1,"delta","/","maxC"]]]]],[13,"v","maxC"],[11,[1,"maxC","===","minC"],[0,[[2,"h","=",[5,"0"]]]],[11,[1,"maxC","===","r"],[0,[[12,"cond",[6,"f32",[[5,"0"]]]],[11,[1,"g","<","b"],[0,[[2,"cond","=",[5,"6"]]]],[0,[[2,"cond","=",[5,"0"]]]]],[2,"h","=",[1,[1,"g","-","b"],"+",[1,"delta","*","cond"]]],[2,"h","/=",[1,[5,"6"],"*","delta"]]]],[11,[1,"maxC","===","g"],[0,[[2,"h","=",[1,[1,"b","-","r"],"+",[1,"delta","*",[5,"2"]]]],[2,"h","/=",[1,[5,"6"],"*","delta"]]]],[11,[1,"maxC","===","b"],[0,[[2,"h","=",[1,[1,"r","-","g"],"+",[1,"delta","*",[5,"4"]]]],[2,"h","/=",[1,[5,"6"],"*","delta"]]]]]]]],[10,[6,"vec3f",["h","s","v"]]]]],"externalNames":["max","min","f32","vec3f"]},
              externals: () => ({max, min, f32, vec3f}),
            }) && $.f)({}))), "rgbToHsv"));

const rgbToYcbcrMatrix = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.const(
  mat3x3f,
  mat3x3f(
    // row 1
    0.299,
    0.587,
    0.114,
    // row 2
    -0.168736,
    -0.331264,
    0.5,
    // row 3
    0.5,
    -0.418688,
    -0.081312
  )
), "rgbToYcbcrMatrix"));
const rgbToYcbcr = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)(
  (($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((rgb) => mul(rgb, rgbToYcbcrMatrix.value)), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"rgb"}],"body":[0,[[10,[6,"mul",["rgb",[7,"rgbToYcbcrMatrix","value"]]]]]],"externalNames":["mul","rgbToYcbcrMatrix"]},
              externals: () => ({mul, rgbToYcbcrMatrix}),
            }) && $.f)({}))
), "rgbToYcbcr"));

const linearToSrgb = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((linear) => {
  return select(
    mul(12.92, linear),
    sub(mul(1.055, pow(linear, vec3f(1 / 2.4))), vec3f(0.055)),
    gt(linear, vec3f(31308e-7))
  );
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"linear"}],"body":[0,[[10,[6,"select",[[6,"mul",[[5,"12.92"],"linear"]],[6,"sub",[[6,"mul",[[5,"1.055"],[6,"pow",["linear",[6,"vec3f",[[1,[5,"1"],"/",[5,"2.4"]]]]]]]],[6,"vec3f",[[5,"0.055"]]]]],[6,"gt",["linear",[6,"vec3f",[[5,"0.0031308"]]]]]]]]]],"externalNames":["select","mul","sub","pow","vec3f","gt"]},
              externals: () => ({select, mul, sub, pow, vec3f, gt}),
            }) && $.f)({}))), "linearToSrgb"));
const srgbToLinear = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((rgb) => {
  return select(
    mul(1 / 12.92, rgb),
    pow(mul(add(rgb, vec3f(0.055)), vec3f(1 / (1 + 0.055))), vec3f(2.4)),
    gt(rgb, vec3f(0.04045))
  );
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"rgb"}],"body":[0,[[10,[6,"select",[[6,"mul",[[1,[5,"1"],"/",[5,"12.92"]],"rgb"]],[6,"pow",[[6,"mul",[[6,"add",["rgb",[6,"vec3f",[[5,"0.055"]]]]],[6,"vec3f",[[1,[5,"1"],"/",[1,[5,"1"],"+",[5,"0.055"]]]]]]],[6,"vec3f",[[5,"2.4"]]]]],[6,"gt",["rgb",[6,"vec3f",[[5,"0.04045"]]]]]]]]]],"externalNames":["select","mul","pow","add","vec3f","gt"]},
              externals: () => ({select, mul, pow, add, vec3f, gt}),
            }) && $.f)({}))), "srgbToLinear"));

const cbrt = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([f32], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((x) => {
  return sign(x) * pow(abs(x), f32(1) / 3);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"x"}],"body":[0,[[10,[1,[6,"sign",["x"]],"*",[6,"pow",[[6,"abs",["x"]],[1,[6,"f32",[[5,"1"]]],"/",[5,"3"]]]]]]]],"externalNames":["sign","pow","abs","f32"]},
              externals: () => ({sign, pow, abs, f32}),
            }) && $.f)({}))), "cbrt"));
const linearRgbToOklab = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((rgb) => {
  const l = 0.4122214708 * rgb.x + 0.5363325363 * rgb.y + 0.0514459929 * rgb.z;
  const m = 0.2119034982 * rgb.x + 0.6806995451 * rgb.y + 0.1073969566 * rgb.z;
  const s = 0.0883024619 * rgb.x + 0.2817188376 * rgb.y + 0.6299787005 * rgb.z;
  const l_ = cbrt(l);
  const m_ = cbrt(m);
  const s_ = cbrt(s);
  return vec3f(
    0.2104542553 * l_ + 0.793617785 * m_ - 0.0040720468 * s_,
    1.9779984951 * l_ - 2.428592205 * m_ + 0.4505937099 * s_,
    0.0259040371 * l_ + 0.7827717662 * m_ - 0.808675766 * s_
  );
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"rgb"}],"body":[0,[[13,"l",[1,[1,[1,[5,"0.4122214708"],"*",[7,"rgb","x"]],"+",[1,[5,"0.5363325363"],"*",[7,"rgb","y"]]],"+",[1,[5,"0.0514459929"],"*",[7,"rgb","z"]]]],[13,"m",[1,[1,[1,[5,"0.2119034982"],"*",[7,"rgb","x"]],"+",[1,[5,"0.6806995451"],"*",[7,"rgb","y"]]],"+",[1,[5,"0.1073969566"],"*",[7,"rgb","z"]]]],[13,"s",[1,[1,[1,[5,"0.0883024619"],"*",[7,"rgb","x"]],"+",[1,[5,"0.2817188376"],"*",[7,"rgb","y"]]],"+",[1,[5,"0.6299787005"],"*",[7,"rgb","z"]]]],[13,"l_",[6,"cbrt",["l"]]],[13,"m_",[6,"cbrt",["m"]]],[13,"s_",[6,"cbrt",["s"]]],[10,[6,"vec3f",[[1,[1,[1,[5,"0.2104542553"],"*","l_"],"+",[1,[5,"0.793617785"],"*","m_"]],"-",[1,[5,"0.0040720468"],"*","s_"]],[1,[1,[1,[5,"1.9779984951"],"*","l_"],"-",[1,[5,"2.428592205"],"*","m_"]],"+",[1,[5,"0.4505937099"],"*","s_"]],[1,[1,[1,[5,"0.0259040371"],"*","l_"],"+",[1,[5,"0.7827717662"],"*","m_"]],"-",[1,[5,"0.808675766"],"*","s_"]]]]]]],"externalNames":["cbrt","vec3f"]},
              externals: () => ({cbrt, vec3f}),
            }) && $.f)({}))), "linearRgbToOklab"));
const oklabToLinearRgb = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((lab) => {
  const l_ = lab.x + 0.3963377774 * lab.y + 0.2158037573 * lab.z;
  const m_ = lab.x - 0.1055613458 * lab.y - 0.0638541728 * lab.z;
  const s_ = lab.x - 0.0894841775 * lab.y - 1.291485548 * lab.z;
  const l = l_ * l_ * l_;
  const m = m_ * m_ * m_;
  const s = s_ * s_ * s_;
  return vec3f(
    4.0767416621 * l - 3.3077115913 * m + 0.2309699292 * s,
    -1.2684380046 * l + 2.6097574011 * m - 0.3413193965 * s,
    -0.0041960863 * l - 0.7034186147 * m + 1.707614701 * s
  );
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"lab"}],"body":[0,[[13,"l_",[1,[1,[7,"lab","x"],"+",[1,[5,"0.3963377774"],"*",[7,"lab","y"]]],"+",[1,[5,"0.2158037573"],"*",[7,"lab","z"]]]],[13,"m_",[1,[1,[7,"lab","x"],"-",[1,[5,"0.1055613458"],"*",[7,"lab","y"]]],"-",[1,[5,"0.0638541728"],"*",[7,"lab","z"]]]],[13,"s_",[1,[1,[7,"lab","x"],"-",[1,[5,"0.0894841775"],"*",[7,"lab","y"]]],"-",[1,[5,"1.291485548"],"*",[7,"lab","z"]]]],[13,"l",[1,[1,"l_","*","l_"],"*","l_"]],[13,"m",[1,[1,"m_","*","m_"],"*","m_"]],[13,"s",[1,[1,"s_","*","s_"],"*","s_"]],[10,[6,"vec3f",[[1,[1,[1,[5,"4.0767416621"],"*","l"],"-",[1,[5,"3.3077115913"],"*","m"]],"+",[1,[5,"0.2309699292"],"*","s"]],[1,[1,[1,[4,"-",[5,"1.2684380046"]],"*","l"],"+",[1,[5,"2.6097574011"],"*","m"]],"-",[1,[5,"0.3413193965"],"*","s"]],[1,[1,[1,[4,"-",[5,"0.0041960863"]],"*","l"],"-",[1,[5,"0.7034186147"],"*","m"]],"+",[1,[5,"1.707614701"],"*","s"]]]]]]],"externalNames":["vec3f"]},
              externals: () => ({vec3f}),
            }) && $.f)({}))), "oklabToLinearRgb"));
const computeMaxSaturation = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([f32, f32], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((a, b) => {
  let k0 = f32(0);
  let k1 = f32(0);
  let k2 = f32(0);
  let k3 = f32(0);
  let k4 = f32(0);
  let wl = f32(0);
  let wm = f32(0);
  let ws = f32(0);
  if (-1.88170328 * a - 0.80936493 * b > 1) {
    k0 = 1.19086277;
    k1 = 1.76576728;
    k2 = 0.59662641;
    k3 = 0.75515197;
    k4 = 0.56771245;
    wl = 4.0767416621;
    wm = -3.3077115913;
    ws = 0.2309699292;
  } else if (1.81444104 * a - 1.19445276 * b > 1) {
    k0 = 0.73956515;
    k1 = -0.45954404;
    k2 = 0.08285427;
    k3 = 0.1254107;
    k4 = 0.14503204;
    wl = -1.2684380046;
    wm = 2.6097574011;
    ws = -0.3413193965;
  } else {
    k0 = 1.35733652;
    k1 = -915799e-8;
    k2 = -1.1513021;
    k3 = -0.50559606;
    k4 = 692167e-8;
    wl = -0.0041960863;
    wm = -0.7034186147;
    ws = 1.707614701;
  }
  const k_l = 0.3963377774 * a + 0.2158037573 * b;
  const k_m = -0.1055613458 * a - 0.0638541728 * b;
  const k_s = -0.0894841775 * a - 1.291485548 * b;
  let S = k0 + k1 * a + k2 * b + k3 * a * a + k4 * a * b;
  {
    const l_ = 1 + S * k_l;
    const m_ = 1 + S * k_m;
    const s_ = 1 + S * k_s;
    const l = l_ * l_ * l_;
    const m = m_ * m_ * m_;
    const s = s_ * s_ * s_;
    const l_dS = 3 * k_l * l_ * l_;
    const m_dS = 3 * k_m * m_ * m_;
    const s_dS = 3 * k_s * s_ * s_;
    const l_dS2 = 6 * k_l * k_l * l_;
    const m_dS2 = 6 * k_m * k_m * m_;
    const s_dS2 = 6 * k_s * k_s * s_;
    const f = wl * l + wm * m + ws * s;
    const f1 = wl * l_dS + wm * m_dS + ws * s_dS;
    const f2 = wl * l_dS2 + wm * m_dS2 + ws * s_dS2;
    S = S - f * f1 / (f1 * f1 - 0.5 * f * f2);
  }
  return S;
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"a"},{"type":"i","name":"b"}],"body":[0,[[12,"k0",[6,"f32",[[5,"0"]]]],[12,"k1",[6,"f32",[[5,"0"]]]],[12,"k2",[6,"f32",[[5,"0"]]]],[12,"k3",[6,"f32",[[5,"0"]]]],[12,"k4",[6,"f32",[[5,"0"]]]],[12,"wl",[6,"f32",[[5,"0"]]]],[12,"wm",[6,"f32",[[5,"0"]]]],[12,"ws",[6,"f32",[[5,"0"]]]],[11,[1,[1,[1,[4,"-",[5,"1.88170328"]],"*","a"],"-",[1,[5,"0.80936493"],"*","b"]],">",[5,"1"]],[0,[[2,"k0","=",[5,"1.19086277"]],[2,"k1","=",[5,"1.76576728"]],[2,"k2","=",[5,"0.59662641"]],[2,"k3","=",[5,"0.75515197"]],[2,"k4","=",[5,"0.56771245"]],[2,"wl","=",[5,"4.0767416621"]],[2,"wm","=",[4,"-",[5,"3.3077115913"]]],[2,"ws","=",[5,"0.2309699292"]]]],[11,[1,[1,[1,[5,"1.81444104"],"*","a"],"-",[1,[5,"1.19445276"],"*","b"]],">",[5,"1"]],[0,[[2,"k0","=",[5,"0.73956515"]],[2,"k1","=",[4,"-",[5,"0.45954404"]]],[2,"k2","=",[5,"0.08285427"]],[2,"k3","=",[5,"0.1254107"]],[2,"k4","=",[5,"0.14503204"]],[2,"wl","=",[4,"-",[5,"1.2684380046"]]],[2,"wm","=",[5,"2.6097574011"]],[2,"ws","=",[4,"-",[5,"0.3413193965"]]]]],[0,[[2,"k0","=",[5,"1.35733652"]],[2,"k1","=",[4,"-",[5,"0.00915799"]]],[2,"k2","=",[4,"-",[5,"1.1513021"]]],[2,"k3","=",[4,"-",[5,"0.50559606"]]],[2,"k4","=",[5,"0.00692167"]],[2,"wl","=",[4,"-",[5,"0.0041960863"]]],[2,"wm","=",[4,"-",[5,"0.7034186147"]]],[2,"ws","=",[5,"1.707614701"]]]]]],[13,"k_l",[1,[1,[5,"0.3963377774"],"*","a"],"+",[1,[5,"0.2158037573"],"*","b"]]],[13,"k_m",[1,[1,[4,"-",[5,"0.1055613458"]],"*","a"],"-",[1,[5,"0.0638541728"],"*","b"]]],[13,"k_s",[1,[1,[4,"-",[5,"0.0894841775"]],"*","a"],"-",[1,[5,"1.291485548"],"*","b"]]],[12,"S",[1,[1,[1,[1,"k0","+",[1,"k1","*","a"]],"+",[1,"k2","*","b"]],"+",[1,[1,"k3","*","a"],"*","a"]],"+",[1,[1,"k4","*","a"],"*","b"]]],[0,[[13,"l_",[1,[5,"1"],"+",[1,"S","*","k_l"]]],[13,"m_",[1,[5,"1"],"+",[1,"S","*","k_m"]]],[13,"s_",[1,[5,"1"],"+",[1,"S","*","k_s"]]],[13,"l",[1,[1,"l_","*","l_"],"*","l_"]],[13,"m",[1,[1,"m_","*","m_"],"*","m_"]],[13,"s",[1,[1,"s_","*","s_"],"*","s_"]],[13,"l_dS",[1,[1,[1,[5,"3"],"*","k_l"],"*","l_"],"*","l_"]],[13,"m_dS",[1,[1,[1,[5,"3"],"*","k_m"],"*","m_"],"*","m_"]],[13,"s_dS",[1,[1,[1,[5,"3"],"*","k_s"],"*","s_"],"*","s_"]],[13,"l_dS2",[1,[1,[1,[5,"6"],"*","k_l"],"*","k_l"],"*","l_"]],[13,"m_dS2",[1,[1,[1,[5,"6"],"*","k_m"],"*","k_m"],"*","m_"]],[13,"s_dS2",[1,[1,[1,[5,"6"],"*","k_s"],"*","k_s"],"*","s_"]],[13,"f",[1,[1,[1,"wl","*","l"],"+",[1,"wm","*","m"]],"+",[1,"ws","*","s"]]],[13,"f1",[1,[1,[1,"wl","*","l_dS"],"+",[1,"wm","*","m_dS"]],"+",[1,"ws","*","s_dS"]]],[13,"f2",[1,[1,[1,"wl","*","l_dS2"],"+",[1,"wm","*","m_dS2"]],"+",[1,"ws","*","s_dS2"]]],[2,"S","=",[1,"S","-",[1,[1,"f","*","f1"],"/",[1,[1,"f1","*","f1"],"-",[1,[1,[5,"0.5"],"*","f"],"*","f2"]]]]]]],[10,"S"]]],"externalNames":["f32"]},
              externals: () => ({f32}),
            }) && $.f)({}))), "computeMaxSaturation"));
const LC = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(struct({
  L: f32,
  C: f32
}), "LC"));
const findCusp = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([f32, f32], LC)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((a, b) => {
  const S_cusp = computeMaxSaturation(a, b);
  const rgb_at_max = oklabToLinearRgb(vec3f(1, S_cusp * a, S_cusp * b));
  const L_cusp = cbrt(1 / max(max(rgb_at_max.x, rgb_at_max.y), rgb_at_max.z));
  const C_cusp = L_cusp * S_cusp;
  return LC({ L: L_cusp, C: C_cusp });
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"a"},{"type":"i","name":"b"}],"body":[0,[[13,"S_cusp",[6,"computeMaxSaturation",["a","b"]]],[13,"rgb_at_max",[6,"oklabToLinearRgb",[[6,"vec3f",[[5,"1"],[1,"S_cusp","*","a"],[1,"S_cusp","*","b"]]]]]],[13,"L_cusp",[6,"cbrt",[[1,[5,"1"],"/",[6,"max",[[6,"max",[[7,"rgb_at_max","x"],[7,"rgb_at_max","y"]]],[7,"rgb_at_max","z"]]]]]]],[13,"C_cusp",[1,"L_cusp","*","S_cusp"]],[10,[6,"LC",[[104,{"L":"L_cusp","C":"C_cusp"}]]]]]],"externalNames":["computeMaxSaturation","oklabToLinearRgb","vec3f","cbrt","max","LC"]},
              externals: () => ({computeMaxSaturation, oklabToLinearRgb, vec3f, cbrt, max, LC}),
            }) && $.f)({}))), "findCusp"));
const findGamutIntersection = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn(
  [f32, f32, f32, f32, f32, LC],
  f32
)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((a, b, L1, C1, L0, cusp) => {
  const FLT_MAX = 340282346e30;
  let t = f32(0);
  if ((L1 - L0) * cusp.C - (cusp.L - L0) * C1 <= 0) {
    t = cusp.C * L0 / (C1 * cusp.L + cusp.C * (L0 - L1));
  } else {
    t = cusp.C * (L0 - 1) / (C1 * (cusp.L - 1) + cusp.C * (L0 - L1));
    {
      const dL = L1 - L0;
      const dC = C1;
      const k_l = 0.3963377774 * a + 0.2158037573 * b;
      const k_m = -0.1055613458 * a - 0.0638541728 * b;
      const k_s = -0.0894841775 * a - 1.291485548 * b;
      const l_dt = dL + dC * k_l;
      const m_dt = dL + dC * k_m;
      const s_dt = dL + dC * k_s;
      {
        const L = L0 * (1 - t) + t * L1;
        const C = t * C1;
        const l_ = L + C * k_l;
        const m_ = L + C * k_m;
        const s_ = L + C * k_s;
        const l = l_ * l_ * l_;
        const m = m_ * m_ * m_;
        const s = s_ * s_ * s_;
        const ldt = 3 * l_dt * l_ * l_;
        const mdt = 3 * m_dt * m_ * m_;
        const sdt = 3 * s_dt * s_ * s_;
        const ldt2 = 6 * l_dt * l_dt * l_;
        const mdt2 = 6 * m_dt * m_dt * m_;
        const sdt2 = 6 * s_dt * s_dt * s_;
        const r = 4.0767416621 * l - 3.3077115913 * m + 0.2309699292 * s - 1;
        const r1 = 4.0767416621 * ldt - 3.3077115913 * mdt + 0.2309699292 * sdt;
        const r2 = 4.0767416621 * ldt2 - 3.3077115913 * mdt2 + 0.2309699292 * sdt2;
        const u_r = r1 / (r1 * r1 - 0.5 * r * r2);
        let t_r = -r * u_r;
        const g = -1.2684380046 * l + 2.6097574011 * m - 0.3413193965 * s - 1;
        const g1 = -1.2684380046 * ldt + 2.6097574011 * mdt - 0.3413193965 * sdt;
        const g2 = -1.2684380046 * ldt2 + 2.6097574011 * mdt2 - 0.3413193965 * sdt2;
        const u_g = g1 / (g1 * g1 - 0.5 * g * g2);
        let t_g = -g * u_g;
        const b2 = -0.0041960863 * l - 0.7034186147 * m + 1.707614701 * s - 1;
        const b1 = -0.0041960863 * ldt - 0.7034186147 * mdt + 1.707614701 * sdt;
        const b22 = -0.0041960863 * ldt2 - 0.7034186147 * mdt2 + 1.707614701 * sdt2;
        const u_b = b1 / (b1 * b1 - 0.5 * b2 * b22);
        let t_b = -b2 * u_b;
        t_r = select(FLT_MAX, t_r, u_r >= 0);
        t_g = select(FLT_MAX, t_g, u_g >= 0);
        t_b = select(FLT_MAX, t_b, u_b >= 0);
        t += min(t_r, min(t_g, t_b));
      }
    }
  }
  return t;
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"a"},{"type":"i","name":"b"},{"type":"i","name":"L1"},{"type":"i","name":"C1"},{"type":"i","name":"L0"},{"type":"i","name":"cusp"}],"body":[0,[[13,"FLT_MAX",[5,"3.40282346e+38"]],[12,"t",[6,"f32",[[5,"0"]]]],[11,[1,[1,[1,[1,"L1","-","L0"],"*",[7,"cusp","C"]],"-",[1,[1,[7,"cusp","L"],"-","L0"],"*","C1"]],"<=",[5,"0"]],[0,[[2,"t","=",[1,[1,[7,"cusp","C"],"*","L0"],"/",[1,[1,"C1","*",[7,"cusp","L"]],"+",[1,[7,"cusp","C"],"*",[1,"L0","-","L1"]]]]]]],[0,[[2,"t","=",[1,[1,[7,"cusp","C"],"*",[1,"L0","-",[5,"1"]]],"/",[1,[1,"C1","*",[1,[7,"cusp","L"],"-",[5,"1"]]],"+",[1,[7,"cusp","C"],"*",[1,"L0","-","L1"]]]]],[0,[[13,"dL",[1,"L1","-","L0"]],[13,"dC","C1"],[13,"k_l",[1,[1,[5,"0.3963377774"],"*","a"],"+",[1,[5,"0.2158037573"],"*","b"]]],[13,"k_m",[1,[1,[4,"-",[5,"0.1055613458"]],"*","a"],"-",[1,[5,"0.0638541728"],"*","b"]]],[13,"k_s",[1,[1,[4,"-",[5,"0.0894841775"]],"*","a"],"-",[1,[5,"1.291485548"],"*","b"]]],[13,"l_dt",[1,"dL","+",[1,"dC","*","k_l"]]],[13,"m_dt",[1,"dL","+",[1,"dC","*","k_m"]]],[13,"s_dt",[1,"dL","+",[1,"dC","*","k_s"]]],[0,[[13,"L",[1,[1,"L0","*",[1,[5,"1"],"-","t"]],"+",[1,"t","*","L1"]]],[13,"C",[1,"t","*","C1"]],[13,"l_",[1,"L","+",[1,"C","*","k_l"]]],[13,"m_",[1,"L","+",[1,"C","*","k_m"]]],[13,"s_",[1,"L","+",[1,"C","*","k_s"]]],[13,"l",[1,[1,"l_","*","l_"],"*","l_"]],[13,"m",[1,[1,"m_","*","m_"],"*","m_"]],[13,"s",[1,[1,"s_","*","s_"],"*","s_"]],[13,"ldt",[1,[1,[1,[5,"3"],"*","l_dt"],"*","l_"],"*","l_"]],[13,"mdt",[1,[1,[1,[5,"3"],"*","m_dt"],"*","m_"],"*","m_"]],[13,"sdt",[1,[1,[1,[5,"3"],"*","s_dt"],"*","s_"],"*","s_"]],[13,"ldt2",[1,[1,[1,[5,"6"],"*","l_dt"],"*","l_dt"],"*","l_"]],[13,"mdt2",[1,[1,[1,[5,"6"],"*","m_dt"],"*","m_dt"],"*","m_"]],[13,"sdt2",[1,[1,[1,[5,"6"],"*","s_dt"],"*","s_dt"],"*","s_"]],[13,"r",[1,[1,[1,[1,[5,"4.0767416621"],"*","l"],"-",[1,[5,"3.3077115913"],"*","m"]],"+",[1,[5,"0.2309699292"],"*","s"]],"-",[5,"1"]]],[13,"r1",[1,[1,[1,[5,"4.0767416621"],"*","ldt"],"-",[1,[5,"3.3077115913"],"*","mdt"]],"+",[1,[5,"0.2309699292"],"*","sdt"]]],[13,"r2",[1,[1,[1,[5,"4.0767416621"],"*","ldt2"],"-",[1,[5,"3.3077115913"],"*","mdt2"]],"+",[1,[5,"0.2309699292"],"*","sdt2"]]],[13,"u_r",[1,"r1","/",[1,[1,"r1","*","r1"],"-",[1,[1,[5,"0.5"],"*","r"],"*","r2"]]]],[12,"t_r",[1,[4,"-","r"],"*","u_r"]],[13,"g",[1,[1,[1,[1,[4,"-",[5,"1.2684380046"]],"*","l"],"+",[1,[5,"2.6097574011"],"*","m"]],"-",[1,[5,"0.3413193965"],"*","s"]],"-",[5,"1"]]],[13,"g1",[1,[1,[1,[4,"-",[5,"1.2684380046"]],"*","ldt"],"+",[1,[5,"2.6097574011"],"*","mdt"]],"-",[1,[5,"0.3413193965"],"*","sdt"]]],[13,"g2",[1,[1,[1,[4,"-",[5,"1.2684380046"]],"*","ldt2"],"+",[1,[5,"2.6097574011"],"*","mdt2"]],"-",[1,[5,"0.3413193965"],"*","sdt2"]]],[13,"u_g",[1,"g1","/",[1,[1,"g1","*","g1"],"-",[1,[1,[5,"0.5"],"*","g"],"*","g2"]]]],[12,"t_g",[1,[4,"-","g"],"*","u_g"]],[13,"b2",[1,[1,[1,[1,[4,"-",[5,"0.0041960863"]],"*","l"],"-",[1,[5,"0.7034186147"],"*","m"]],"+",[1,[5,"1.707614701"],"*","s"]],"-",[5,"1"]]],[13,"b1",[1,[1,[1,[4,"-",[5,"0.0041960863"]],"*","ldt"],"-",[1,[5,"0.7034186147"],"*","mdt"]],"+",[1,[5,"1.707614701"],"*","sdt"]]],[13,"b22",[1,[1,[1,[4,"-",[5,"0.0041960863"]],"*","ldt2"],"-",[1,[5,"0.7034186147"],"*","mdt2"]],"+",[1,[5,"1.707614701"],"*","sdt2"]]],[13,"u_b",[1,"b1","/",[1,[1,"b1","*","b1"],"-",[1,[1,[5,"0.5"],"*","b2"],"*","b22"]]]],[12,"t_b",[1,[4,"-","b2"],"*","u_b"]],[2,"t_r","=",[6,"select",["FLT_MAX","t_r",[1,"u_r",">=",[5,"0"]]]]],[2,"t_g","=",[6,"select",["FLT_MAX","t_g",[1,"u_g",">=",[5,"0"]]]]],[2,"t_b","=",[6,"select",["FLT_MAX","t_b",[1,"u_b",">=",[5,"0"]]]]],[2,"t","+=",[6,"min",["t_r",[6,"min",["t_g","t_b"]]]]]]]]]]]],[10,"t"]]],"externalNames":["f32","select","min"]},
              externals: () => ({f32, select, min}),
            }) && $.f)({}))), "findGamutIntersection"));
const gamutClipPreserveChroma = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((lab) => {
  const L = lab.x;
  const eps = 1e-5;
  const C = max(eps, length(lab.yz));
  const a_ = lab.y / C;
  const b_ = lab.z / C;
  const L0 = clamp(L, 0, 1);
  const cusp = findCusp(a_, b_);
  const t = clamp(findGamutIntersection(a_, b_, L, C, L0, cusp), 0, 1);
  const L_clipped = mix(L0, L, t);
  const C_clipped = t * C;
  return vec3f(L_clipped, C_clipped * a_, C_clipped * b_);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"lab"}],"body":[0,[[13,"L",[7,"lab","x"]],[13,"eps",[5,"0.00001"]],[13,"C",[6,"max",["eps",[6,"length",[[7,"lab","yz"]]]]]],[13,"a_",[1,[7,"lab","y"],"/","C"]],[13,"b_",[1,[7,"lab","z"],"/","C"]],[13,"L0",[6,"clamp",["L",[5,"0"],[5,"1"]]]],[13,"cusp",[6,"findCusp",["a_","b_"]]],[13,"t",[6,"clamp",[[6,"findGamutIntersection",["a_","b_","L","C","L0","cusp"]],[5,"0"],[5,"1"]]]],[13,"L_clipped",[6,"mix",["L0","L","t"]]],[13,"C_clipped",[1,"t","*","C"]],[10,[6,"vec3f",["L_clipped",[1,"C_clipped","*","a_"],[1,"C_clipped","*","b_"]]]]]],"externalNames":["max","length","clamp","findCusp","findGamutIntersection","mix","vec3f"]},
              externals: () => ({max, length, clamp, findCusp, findGamutIntersection, mix, vec3f}),
            }) && $.f)({}))), "gamutClipPreserveChroma"));
const oklabGamutClipAlphaAccess = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu["~unstable"].accessor(f32, 0.2), "oklabGamutClipAlphaAccess"));
const gamutClipAdaptiveL05 = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((lab) => {
  const alpha = oklabGamutClipAlphaAccess.value;
  const L = lab.x;
  const eps = 1e-5;
  const C = max(eps, length(lab.yz));
  const a_ = lab.y / C;
  const b_ = lab.z / C;
  const Ld = L - 0.5;
  const e1 = 0.5 + abs(Ld) + alpha * C;
  const L0 = 0.5 * (1 + sign(Ld) * (e1 - sqrt(max(0, e1 * e1 - 2 * abs(Ld)))));
  const cusp = findCusp(a_, b_);
  const t = clamp(findGamutIntersection(a_, b_, L, C, L0, cusp), 0, 1);
  const L_clipped = mix(L0, L, t);
  const C_clipped = t * C;
  return vec3f(L_clipped, C_clipped * a_, C_clipped * b_);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"lab"}],"body":[0,[[13,"alpha",[7,"oklabGamutClipAlphaAccess","value"]],[13,"L",[7,"lab","x"]],[13,"eps",[5,"0.00001"]],[13,"C",[6,"max",["eps",[6,"length",[[7,"lab","yz"]]]]]],[13,"a_",[1,[7,"lab","y"],"/","C"]],[13,"b_",[1,[7,"lab","z"],"/","C"]],[13,"Ld",[1,"L","-",[5,"0.5"]]],[13,"e1",[1,[1,[5,"0.5"],"+",[6,"abs",["Ld"]]],"+",[1,"alpha","*","C"]]],[13,"L0",[1,[5,"0.5"],"*",[1,[5,"1"],"+",[1,[6,"sign",["Ld"]],"*",[1,"e1","-",[6,"sqrt",[[6,"max",[[5,"0"],[1,[1,"e1","*","e1"],"-",[1,[5,"2"],"*",[6,"abs",["Ld"]]]]]]]]]]]]],[13,"cusp",[6,"findCusp",["a_","b_"]]],[13,"t",[6,"clamp",[[6,"findGamutIntersection",["a_","b_","L","C","L0","cusp"]],[5,"0"],[5,"1"]]]],[13,"L_clipped",[6,"mix",["L0","L","t"]]],[13,"C_clipped",[1,"t","*","C"]],[10,[6,"vec3f",["L_clipped",[1,"C_clipped","*","a_"],[1,"C_clipped","*","b_"]]]]]],"externalNames":["oklabGamutClipAlphaAccess","max","length","abs","sign","sqrt","findCusp","clamp","findGamutIntersection","mix","vec3f"]},
              externals: () => ({oklabGamutClipAlphaAccess, max, length, abs, sign, sqrt, findCusp, clamp, findGamutIntersection, mix, vec3f}),
            }) && $.f)({}))), "gamutClipAdaptiveL05"));
const gamutClipAdaptiveL0cusp = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((lab) => {
  const alpha = oklabGamutClipAlphaAccess.value;
  const L = lab.x;
  const eps = 1e-5;
  const C = max(eps, length(lab.yz));
  const a_ = lab.y / C;
  const b_ = lab.z / C;
  const cusp = findCusp(a_, b_);
  const Ld = L - cusp.L;
  const k = 2 * select(cusp.L, 1 - cusp.L, Ld > 0);
  const e1 = 0.5 * k + abs(Ld) + alpha * C / k;
  const L0 = cusp.L + 0.5 * (sign(Ld) * (e1 - sqrt(max(0, e1 * e1 - 2 * k * abs(Ld)))));
  const t = clamp(findGamutIntersection(a_, b_, L, C, L0, cusp), 0, 1);
  const L_clipped = mix(L0, L, t);
  const C_clipped = t * C;
  return vec3f(L_clipped, C_clipped * a_, C_clipped * b_);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"lab"}],"body":[0,[[13,"alpha",[7,"oklabGamutClipAlphaAccess","value"]],[13,"L",[7,"lab","x"]],[13,"eps",[5,"0.00001"]],[13,"C",[6,"max",["eps",[6,"length",[[7,"lab","yz"]]]]]],[13,"a_",[1,[7,"lab","y"],"/","C"]],[13,"b_",[1,[7,"lab","z"],"/","C"]],[13,"cusp",[6,"findCusp",["a_","b_"]]],[13,"Ld",[1,"L","-",[7,"cusp","L"]]],[13,"k",[1,[5,"2"],"*",[6,"select",[[7,"cusp","L"],[1,[5,"1"],"-",[7,"cusp","L"]],[1,"Ld",">",[5,"0"]]]]]],[13,"e1",[1,[1,[1,[5,"0.5"],"*","k"],"+",[6,"abs",["Ld"]]],"+",[1,[1,"alpha","*","C"],"/","k"]]],[13,"L0",[1,[7,"cusp","L"],"+",[1,[5,"0.5"],"*",[1,[6,"sign",["Ld"]],"*",[1,"e1","-",[6,"sqrt",[[6,"max",[[5,"0"],[1,[1,"e1","*","e1"],"-",[1,[1,[5,"2"],"*","k"],"*",[6,"abs",["Ld"]]]]]]]]]]]]],[13,"t",[6,"clamp",[[6,"findGamutIntersection",["a_","b_","L","C","L0","cusp"]],[5,"0"],[5,"1"]]]],[13,"L_clipped",[6,"mix",["L0","L","t"]]],[13,"C_clipped",[1,"t","*","C"]],[10,[6,"vec3f",["L_clipped",[1,"C_clipped","*","a_"],[1,"C_clipped","*","b_"]]]]]],"externalNames":["oklabGamutClipAlphaAccess","max","length","findCusp","select","abs","sign","sqrt","clamp","findGamutIntersection","mix","vec3f"]},
              externals: () => ({oklabGamutClipAlphaAccess, max, length, findCusp, select, abs, sign, sqrt, clamp, findGamutIntersection, mix, vec3f}),
            }) && $.f)({}))), "gamutClipAdaptiveL0cusp"));
const oklabGamutClipSlot = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.slot(gamutClipAdaptiveL05), "oklabGamutClipSlot"));
const oklabGamutClip = {
  preserveChroma: gamutClipPreserveChroma,
  adaptiveL05: gamutClipAdaptiveL05,
  adaptiveL0Cusp: gamutClipAdaptiveL0cusp
};
const oklabToRgb = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((lab) => {
  return linearToSrgb(oklabToLinearRgb(oklabGamutClipSlot.value(lab)));
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"lab"}],"body":[0,[[10,[6,"linearToSrgb",[[6,"oklabToLinearRgb",[[6,[7,"oklabGamutClipSlot","value"],["lab"]]]]]]]]],"externalNames":["linearToSrgb","oklabToLinearRgb","oklabGamutClipSlot"]},
              externals: () => ({linearToSrgb, oklabToLinearRgb, oklabGamutClipSlot}),
            }) && $.f)({}))), "oklabToRgb"));
const rgbToOklab = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f], vec3f)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((rgb) => {
  return linearRgbToOklab(srgbToLinear(rgb));
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"rgb"}],"body":[0,[[10,[6,"linearRgbToOklab",[[6,"srgbToLinear",["rgb"]]]]]]],"externalNames":["linearRgbToOklab","srgbToLinear"]},
              externals: () => ({linearRgbToOklab, srgbToLinear}),
            }) && $.f)({}))), "rgbToOklab"));

export { hsvToRgb, linearRgbToOklab, linearToSrgb, oklabGamutClip, oklabGamutClipAlphaAccess, oklabGamutClipSlot, oklabToLinearRgb, oklabToRgb, rgbToHsv, rgbToOklab, rgbToYcbcr, rgbToYcbcrMatrix, srgbToLinear };
