// @typegpu/sdf/index.mjs
import tgpu from 'typegpu';
import * as d from 'typegpu/data';
import { vec2f, f32, vec3f } from 'typegpu/data';
import * as std from 'typegpu/std';
import { length, sub, abs, max, min, add, dot, sqrt, sign, pow, clamp, acos, cos, mul, sin, saturate } from 'typegpu/std';

const sdDisk = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec2f, f32], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, radius) => {
  return length(point) - radius;
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"radius"}],"body":[0,[[10,[1,[6,"length",["point"]],"-","radius"]]]],"externalNames":["length"]},
              externals: () => ({length}),
            }) && $.f)({}))), "sdDisk"));
const sdBox2d = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec2f, vec2f], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, size) => {
  const d = sub(abs(point), size);
  return length(max(d, vec2f(0))) + min(max(d.x, d.y), 0);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"size"}],"body":[0,[[13,"d",[6,"sub",[[6,"abs",["point"]],"size"]]],[10,[1,[6,"length",[[6,"max",["d",[6,"vec2f",[[5,"0"]]]]]]],"+",[6,"min",[[6,"max",[[7,"d","x"],[7,"d","y"]]],[5,"0"]]]]]]],"externalNames":["sub","abs","length","max","vec2f","min"]},
              externals: () => ({sub, abs, length, max, vec2f, min}),
            }) && $.f)({}))), "sdBox2d"));
const sdRoundedBox2d = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec2f, vec2f, f32], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, size, cornerRadius) => {
  const d = add(sub(abs(point), size), vec2f(cornerRadius));
  return length(max(d, vec2f(0))) + min(max(d.x, d.y), 0) - cornerRadius;
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"size"},{"type":"i","name":"cornerRadius"}],"body":[0,[[13,"d",[6,"add",[[6,"sub",[[6,"abs",["point"]],"size"]],[6,"vec2f",["cornerRadius"]]]]],[10,[1,[1,[6,"length",[[6,"max",["d",[6,"vec2f",[[5,"0"]]]]]]],"+",[6,"min",[[6,"max",[[7,"d","x"],[7,"d","y"]]],[5,"0"]]]],"-","cornerRadius"]]]],"externalNames":["add","sub","abs","vec2f","length","max","min"]},
              externals: () => ({add, sub, abs, vec2f, length, max, min}),
            }) && $.f)({}))), "sdRoundedBox2d"));
const sdLine = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec2f, vec2f, vec2f], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, A, B) => {
  const pa = sub(point, A);
  const ba = sub(B, A);
  const h = max(0, min(1, dot(pa, ba) / dot(ba, ba)));
  return length(sub(pa, ba.mul(h)));
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"A"},{"type":"i","name":"B"}],"body":[0,[[13,"pa",[6,"sub",["point","A"]]],[13,"ba",[6,"sub",["B","A"]]],[13,"h",[6,"max",[[5,"0"],[6,"min",[[5,"1"],[1,[6,"dot",["pa","ba"]],"/",[6,"dot",["ba","ba"]]]]]]]],[10,[6,"length",[[6,"sub",["pa",[6,[7,"ba","mul"],["h"]]]]]]]]],"externalNames":["sub","max","min","dot","length"]},
              externals: () => ({sub, max, min, dot, length}),
            }) && $.f)({}))), "sdLine"));
const dot2 = (($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((v) => {
  "use gpu";
  return dot(v, v);
}), {
              v: 1,
              name: "dot2",
              ast: {"params":[{"type":"i","name":"v"}],"body":[0,[[10,[6,"dot",["v","v"]]]]],"externalNames":["dot"]},
              externals: () => ({dot}),
            }) && $.f)({}));
const sdBezier = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec2f, vec2f, vec2f, vec2f], f32)(
  (($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, A, B, C) => {
    const a = B.sub(A);
    const b = A.sub(B.mul(2)).add(C);
    const c = a.mul(f32(2));
    const d = A.sub(point);
    const dotB = max(dot(b, b), 1e-4);
    const kk = 1 / dotB;
    const kx = kk * dot(a, b);
    const ky = kk * (f32(2) * dot(a, a) + dot(d, b)) / 3;
    const kz = kk * dot(d, a);
    let res = f32(0);
    const p = ky - kx * kx;
    const p3 = p * p * p;
    const q = kx * (2 * kx * kx - 3 * ky) + kz;
    let h = q * q + 4 * p3;
    if (h >= 0) {
      h = sqrt(h);
      const x = vec2f(h, -h).sub(q).mul(0.5);
      const uv = sign(x).mul(pow(abs(x), vec2f(1 / 3)));
      const t = clamp(uv.x + uv.y - kx, 0, 1);
      res = dot2(d.add(c.add(b.mul(t)).mul(t)));
    } else {
      const z = sqrt(-p);
      const v = acos(q / (p * z * 2)) / 3;
      const m = cos(v);
      const n = mul(sin(v), 1.732050808);
      const t = saturate(
        vec3f(m + m, -n - m, n - m).mul(z).sub(kx)
      );
      res = min(
        dot2(d.add(c.add(b.mul(t.x)).mul(t.x))),
        dot2(d.add(c.add(b.mul(t.y)).mul(t.y)))
      );
    }
    return sqrt(res);
  }), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"A"},{"type":"i","name":"B"},{"type":"i","name":"C"}],"body":[0,[[13,"a",[6,[7,"B","sub"],["A"]]],[13,"b",[6,[7,[6,[7,"A","sub"],[[6,[7,"B","mul"],[[5,"2"]]]]],"add"],["C"]]],[13,"c",[6,[7,"a","mul"],[[6,"f32",[[5,"2"]]]]]],[13,"d",[6,[7,"A","sub"],["point"]]],[13,"dotB",[6,"max",[[6,"dot",["b","b"]],[5,"0.0001"]]]],[13,"kk",[1,[5,"1"],"/","dotB"]],[13,"kx",[1,"kk","*",[6,"dot",["a","b"]]]],[13,"ky",[1,[1,"kk","*",[1,[1,[6,"f32",[[5,"2"]]],"*",[6,"dot",["a","a"]]],"+",[6,"dot",["d","b"]]]],"/",[5,"3"]]],[13,"kz",[1,"kk","*",[6,"dot",["d","a"]]]],[12,"res",[6,"f32",[[5,"0"]]]],[13,"p",[1,"ky","-",[1,"kx","*","kx"]]],[13,"p3",[1,[1,"p","*","p"],"*","p"]],[13,"q",[1,[1,"kx","*",[1,[1,[1,[5,"2"],"*","kx"],"*","kx"],"-",[1,[5,"3"],"*","ky"]]],"+","kz"]],[12,"h",[1,[1,"q","*","q"],"+",[1,[5,"4"],"*","p3"]]],[11,[1,"h",">=",[5,"0"]],[0,[[2,"h","=",[6,"sqrt",["h"]]],[13,"x",[6,[7,[6,[7,[6,"vec2f",["h",[4,"-","h"]]],"sub"],["q"]],"mul"],[[5,"0.5"]]]],[13,"uv",[6,[7,[6,"sign",["x"]],"mul"],[[6,"pow",[[6,"abs",["x"]],[6,"vec2f",[[1,[5,"1"],"/",[5,"3"]]]]]]]]],[13,"t",[6,"clamp",[[1,[1,[7,"uv","x"],"+",[7,"uv","y"]],"-","kx"],[5,"0"],[5,"1"]]]],[2,"res","=",[6,"dot2",[[6,[7,"d","add"],[[6,[7,[6,[7,"c","add"],[[6,[7,"b","mul"],["t"]]]],"mul"],["t"]]]]]]]]],[0,[[13,"z",[6,"sqrt",[[4,"-","p"]]]],[13,"v",[1,[6,"acos",[[1,"q","/",[1,[1,"p","*","z"],"*",[5,"2"]]]]],"/",[5,"3"]]],[13,"m",[6,"cos",["v"]]],[13,"n",[6,"mul",[[6,"sin",["v"]],[5,"1.732050808"]]]],[13,"t",[6,"saturate",[[6,[7,[6,[7,[6,"vec3f",[[1,"m","+","m"],[1,[4,"-","n"],"-","m"],[1,"n","-","m"]]],"mul"],["z"]],"sub"],["kx"]]]]],[2,"res","=",[6,"min",[[6,"dot2",[[6,[7,"d","add"],[[6,[7,[6,[7,"c","add"],[[6,[7,"b","mul"],[[7,"t","x"]]]]],"mul"],[[7,"t","x"]]]]]]],[6,"dot2",[[6,[7,"d","add"],[[6,[7,[6,[7,"c","add"],[[6,[7,"b","mul"],[[7,"t","y"]]]]],"mul"],[[7,"t","y"]]]]]]]]]]]]],[10,[6,"sqrt",["res"]]]]],"externalNames":["f32","max","dot","sqrt","vec2f","sign","pow","abs","clamp","dot2","acos","cos","mul","sin","saturate","vec3f","min"]},
              externals: () => ({f32, max, dot, sqrt, vec2f, sign, pow, abs, clamp, dot2, acos, cos, mul, sin, saturate, vec3f, min}),
            }) && $.f)({}))
), "sdBezier"));
const cross = (($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((a, b) => {
  "use gpu";
  return a.x * b.y - a.y * b.x;
}), {
              v: 1,
              name: "cross",
              ast: {"params":[{"type":"i","name":"a"},{"type":"i","name":"b"}],"body":[0,[[10,[1,[1,[7,"a","x"],"*",[7,"b","y"]],"-",[1,[7,"a","y"],"*",[7,"b","x"]]]]]],"externalNames":[]},
              externals: () => ({}),
            }) && $.f)({}));
const sdBezierApprox = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn(
  [vec2f, vec2f, vec2f, vec2f],
  f32
)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, A, B, C) => {
  const i = A.sub(C);
  const j = C.sub(B);
  const k = B.sub(A);
  const w = j.sub(k);
  const v0 = A.sub(point);
  const v1 = B.sub(point);
  const v2 = C.sub(point);
  const x = cross(v0, v2);
  const y = cross(v1, v0);
  const z = cross(v2, v1);
  const s = j.mul(y).add(k.mul(z)).mul(2).sub(i.mul(x));
  const r = (y * z - x * x * 0.25) / dot2(s);
  const t = saturate((0.5 * x + y + r * dot(s, w)) / (x + y + z));
  const d = v0.add(k.add(k).add(w.mul(t)).mul(t));
  return length(d);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"A"},{"type":"i","name":"B"},{"type":"i","name":"C"}],"body":[0,[[13,"i",[6,[7,"A","sub"],["C"]]],[13,"j",[6,[7,"C","sub"],["B"]]],[13,"k",[6,[7,"B","sub"],["A"]]],[13,"w",[6,[7,"j","sub"],["k"]]],[13,"v0",[6,[7,"A","sub"],["point"]]],[13,"v1",[6,[7,"B","sub"],["point"]]],[13,"v2",[6,[7,"C","sub"],["point"]]],[13,"x",[6,"cross",["v0","v2"]]],[13,"y",[6,"cross",["v1","v0"]]],[13,"z",[6,"cross",["v2","v1"]]],[13,"s",[6,[7,[6,[7,[6,[7,[6,[7,"j","mul"],["y"]],"add"],[[6,[7,"k","mul"],["z"]]]],"mul"],[[5,"2"]]],"sub"],[[6,[7,"i","mul"],["x"]]]]],[13,"r",[1,[1,[1,"y","*","z"],"-",[1,[1,"x","*","x"],"*",[5,"0.25"]]],"/",[6,"dot2",["s"]]]],[13,"t",[6,"saturate",[[1,[1,[1,[1,[5,"0.5"],"*","x"],"+","y"],"+",[1,"r","*",[6,"dot",["s","w"]]]],"/",[1,[1,"x","+","y"],"+","z"]]]]],[13,"d",[6,[7,"v0","add"],[[6,[7,[6,[7,[6,[7,"k","add"],["k"]],"add"],[[6,[7,"w","mul"],["t"]]]],"mul"],["t"]]]]],[10,[6,"length",["d"]]]]],"externalNames":["cross","dot2","saturate","dot","length"]},
              externals: () => ({cross, dot2, saturate, dot, length}),
            }) && $.f)({}))), "sdBezierApprox"));
const sdPie = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec2f, vec2f, f32], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, sc, radius) => {
  const p_w = vec2f(point);
  p_w.x = abs(point.x);
  const l = length(p_w) - radius;
  const m = length(p_w.sub(sc.mul(clamp(dot(p_w, sc), 0, radius))));
  return max(l, m * sign(sc.y * p_w.x - sc.x * p_w.y));
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"sc"},{"type":"i","name":"radius"}],"body":[0,[[13,"p_w",[6,"vec2f",["point"]]],[2,[7,"p_w","x"],"=",[6,"abs",[[7,"point","x"]]]],[13,"l",[1,[6,"length",["p_w"]],"-","radius"]],[13,"m",[6,"length",[[6,[7,"p_w","sub"],[[6,[7,"sc","mul"],[[6,"clamp",[[6,"dot",["p_w","sc"]],[5,"0"],"radius"]]]]]]]]],[10,[6,"max",["l",[1,"m","*",[6,"sign",[[1,[1,[7,"sc","y"],"*",[7,"p_w","x"]],"-",[1,[7,"sc","x"],"*",[7,"p_w","y"]]]]]]]]]]],"externalNames":["vec2f","abs","length","clamp","dot","max","sign"]},
              externals: () => ({vec2f, abs, length, clamp, dot, max, sign}),
            }) && $.f)({}))), "sdPie"));

const sdSphere = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f, f32], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, radius) => {
  return length(point) - radius;
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"radius"}],"body":[0,[[10,[1,[6,"length",["point"]],"-","radius"]]]],"externalNames":["length"]},
              externals: () => ({length}),
            }) && $.f)({}))), "sdSphere"));
const sdBox3d = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f, vec3f], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, size) => {
  const d = sub(abs(point), size);
  return length(max(d, vec3f(0))) + min(max(max(d.x, d.y), d.z), 0);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"size"}],"body":[0,[[13,"d",[6,"sub",[[6,"abs",["point"]],"size"]]],[10,[1,[6,"length",[[6,"max",["d",[6,"vec3f",[[5,"0"]]]]]]],"+",[6,"min",[[6,"max",[[6,"max",[[7,"d","x"],[7,"d","y"]]],[7,"d","z"]]],[5,"0"]]]]]]],"externalNames":["sub","abs","length","max","vec3f","min"]},
              externals: () => ({sub, abs, length, max, vec3f, min}),
            }) && $.f)({}))), "sdBox3d"));
const sdRoundedBox3d = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f, vec3f, f32], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, size, cornerRadius) => {
  const d = add(sub(abs(point), size), vec3f(cornerRadius));
  return length(max(d, vec3f(0))) + min(max(max(d.x, d.y), d.z), 0) - cornerRadius;
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"size"},{"type":"i","name":"cornerRadius"}],"body":[0,[[13,"d",[6,"add",[[6,"sub",[[6,"abs",["point"]],"size"]],[6,"vec3f",["cornerRadius"]]]]],[10,[1,[1,[6,"length",[[6,"max",["d",[6,"vec3f",[[5,"0"]]]]]]],"+",[6,"min",[[6,"max",[[6,"max",[[7,"d","x"],[7,"d","y"]]],[7,"d","z"]]],[5,"0"]]]],"-","cornerRadius"]]]],"externalNames":["add","sub","abs","vec3f","length","max","min"]},
              externals: () => ({add, sub, abs, vec3f, length, max, min}),
            }) && $.f)({}))), "sdRoundedBox3d"));
const sdBoxFrame3d = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f, vec3f, f32], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, size, thickness) => {
  const p1 = sub(abs(point), size);
  const q = sub(abs(add(p1, thickness)), vec3f(thickness));
  const d1 = length(max(vec3f(p1.x, q.y, q.z), vec3f(0))) + min(max(p1.x, max(q.y, q.z)), 0);
  const d2 = length(max(vec3f(q.x, p1.y, q.z), vec3f(0))) + min(max(q.x, max(p1.y, q.z)), 0);
  const d3 = length(max(vec3f(q.x, q.y, p1.z), vec3f(0))) + min(max(q.x, max(q.y, p1.z)), 0);
  return min(min(d1, d2), d3);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"size"},{"type":"i","name":"thickness"}],"body":[0,[[13,"p1",[6,"sub",[[6,"abs",["point"]],"size"]]],[13,"q",[6,"sub",[[6,"abs",[[6,"add",["p1","thickness"]]]],[6,"vec3f",["thickness"]]]]],[13,"d1",[1,[6,"length",[[6,"max",[[6,"vec3f",[[7,"p1","x"],[7,"q","y"],[7,"q","z"]]],[6,"vec3f",[[5,"0"]]]]]]],"+",[6,"min",[[6,"max",[[7,"p1","x"],[6,"max",[[7,"q","y"],[7,"q","z"]]]]],[5,"0"]]]]],[13,"d2",[1,[6,"length",[[6,"max",[[6,"vec3f",[[7,"q","x"],[7,"p1","y"],[7,"q","z"]]],[6,"vec3f",[[5,"0"]]]]]]],"+",[6,"min",[[6,"max",[[7,"q","x"],[6,"max",[[7,"p1","y"],[7,"q","z"]]]]],[5,"0"]]]]],[13,"d3",[1,[6,"length",[[6,"max",[[6,"vec3f",[[7,"q","x"],[7,"q","y"],[7,"p1","z"]]],[6,"vec3f",[[5,"0"]]]]]]],"+",[6,"min",[[6,"max",[[7,"q","x"],[6,"max",[[7,"q","y"],[7,"p1","z"]]]]],[5,"0"]]]]],[10,[6,"min",[[6,"min",["d1","d2"]],"d3"]]]]],"externalNames":["sub","abs","add","vec3f","length","max","min"]},
              externals: () => ({sub, abs, add, vec3f, length, max, min}),
            }) && $.f)({}))), "sdBoxFrame3d"));
const sdLine3d = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f, vec3f, vec3f], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, A, B) => {
  const pa = sub(point, A);
  const ba = sub(B, A);
  const h = max(0, min(1, dot(pa, ba) / dot(ba, ba)));
  return length(sub(pa, ba.mul(h)));
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"A"},{"type":"i","name":"B"}],"body":[0,[[13,"pa",[6,"sub",["point","A"]]],[13,"ba",[6,"sub",["B","A"]]],[13,"h",[6,"max",[[5,"0"],[6,"min",[[5,"1"],[1,[6,"dot",["pa","ba"]],"/",[6,"dot",["ba","ba"]]]]]]]],[10,[6,"length",[[6,"sub",["pa",[6,[7,"ba","mul"],["h"]]]]]]]]],"externalNames":["sub","max","min","dot","length"]},
              externals: () => ({sub, max, min, dot, length}),
            }) && $.f)({}))), "sdLine3d"));
const sdPlane = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f, vec3f, f32], f32)(
  (($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, normal, height) => {
    return dot(point, normal) + height;
  }), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"normal"},{"type":"i","name":"height"}],"body":[0,[[10,[1,[6,"dot",["point","normal"]],"+","height"]]]],"externalNames":["dot"]},
              externals: () => ({dot}),
            }) && $.f)({}))
), "sdPlane"));
const sdCapsule = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([vec3f, vec3f, vec3f, f32], f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, A, B, radius) => {
  const pa = sub(point, A);
  const ba = sub(B, A);
  const h = saturate(dot(pa, ba) / dot(ba, ba));
  return length(sub(pa, ba.mul(h))) - radius;
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"A"},{"type":"i","name":"B"},{"type":"i","name":"radius"}],"body":[0,[[13,"pa",[6,"sub",["point","A"]]],[13,"ba",[6,"sub",["B","A"]]],[13,"h",[6,"saturate",[[1,[6,"dot",["pa","ba"]],"/",[6,"dot",["ba","ba"]]]]]],[10,[1,[6,"length",[[6,"sub",["pa",[6,[7,"ba","mul"],["h"]]]]]],"-","radius"]]]],"externalNames":["sub","saturate","dot","length"]},
              externals: () => ({sub, saturate, dot, length}),
            }) && $.f)({}))), "sdCapsule"));

const opUnion = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([d.f32, d.f32], d.f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((d1, d2) => std.min(d1, d2)), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"d1"},{"type":"i","name":"d2"}],"body":[0,[[10,[6,[7,"std","min"],["d1","d2"]]]]],"externalNames":["std"]},
              externals: () => ({std}),
            }) && $.f)({}))), "opUnion"));
const opSmoothUnion = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([d.f32, d.f32, d.f32], d.f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((d1, d2, k) => {
  const h = std.max(k - std.abs(d1 - d2), 0) / k;
  return std.min(d1, d2) - h * h * k * (1 / 4);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"d1"},{"type":"i","name":"d2"},{"type":"i","name":"k"}],"body":[0,[[13,"h",[1,[6,[7,"std","max"],[[1,"k","-",[6,[7,"std","abs"],[[1,"d1","-","d2"]]]],[5,"0"]]],"/","k"]],[10,[1,[6,[7,"std","min"],["d1","d2"]],"-",[1,[1,[1,"h","*","h"],"*","k"],"*",[1,[5,"1"],"/",[5,"4"]]]]]]],"externalNames":["std"]},
              externals: () => ({std}),
            }) && $.f)({}))), "opSmoothUnion"));
const opSmoothDifference = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([d.f32, d.f32, d.f32], d.f32)((($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((d1, d2, k) => {
  const h = std.max(k - std.abs(d1 + d2), 0) / k;
  return std.max(-d2, d1) + h * h * k * (1 / 4);
}), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"d1"},{"type":"i","name":"d2"},{"type":"i","name":"k"}],"body":[0,[[13,"h",[1,[6,[7,"std","max"],[[1,"k","-",[6,[7,"std","abs"],[[1,"d1","+","d2"]]]],[5,"0"]]],"/","k"]],[10,[1,[6,[7,"std","max"],[[4,"-","d2"],"d1"]],"+",[1,[1,[1,"h","*","h"],"*","k"],"*",[1,[5,"1"],"/",[5,"4"]]]]]]],"externalNames":["std"]},
              externals: () => ({std}),
            }) && $.f)({}))), "opSmoothDifference"));
const opExtrudeZ = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([d.vec3f, d.f32, d.f32], d.f32)(
  (($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, dd, halfHeight) => {
    const w = d.vec2f(dd, std.abs(point.z) - halfHeight);
    return std.min(std.max(w.x, w.y), 0) + std.length(std.max(w, d.vec2f()));
  }), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"dd"},{"type":"i","name":"halfHeight"}],"body":[0,[[13,"w",[6,[7,"d","vec2f"],["dd",[1,[6,[7,"std","abs"],[[7,"point","z"]]],"-","halfHeight"]]]],[10,[1,[6,[7,"std","min"],[[6,[7,"std","max"],[[7,"w","x"],[7,"w","y"]]],[5,"0"]]],"+",[6,[7,"std","length"],[[6,[7,"std","max"],["w",[6,[7,"d","vec2f"],[]]]]]]]]]],"externalNames":["d","std"]},
              externals: () => ({d, std}),
            }) && $.f)({}))
), "opExtrudeZ"));
const opExtrudeX = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([d.vec3f, d.f32, d.f32], d.f32)(
  (($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, dd, halfHeight) => {
    const w = d.vec2f(dd, std.abs(point.x) - halfHeight);
    return std.min(std.max(w.x, w.y), 0) + std.length(std.max(w, d.vec2f()));
  }), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"dd"},{"type":"i","name":"halfHeight"}],"body":[0,[[13,"w",[6,[7,"d","vec2f"],["dd",[1,[6,[7,"std","abs"],[[7,"point","x"]]],"-","halfHeight"]]]],[10,[1,[6,[7,"std","min"],[[6,[7,"std","max"],[[7,"w","x"],[7,"w","y"]]],[5,"0"]]],"+",[6,[7,"std","length"],[[6,[7,"std","max"],["w",[6,[7,"d","vec2f"],[]]]]]]]]]],"externalNames":["d","std"]},
              externals: () => ({d, std}),
            }) && $.f)({}))
), "opExtrudeX"));
const opExtrudeY = ((globalThis.__TYPEGPU_AUTONAME__ ?? (a => a))(tgpu.fn([d.vec3f, d.f32, d.f32], d.f32)(
  (($ => (globalThis.__TYPEGPU_META__ ??= new WeakMap()).set($.f = ((point, dd, halfHeight) => {
    const w = d.vec2f(dd, std.abs(point.y) - halfHeight);
    return std.min(std.max(w.x, w.y), 0) + std.length(std.max(w, d.vec2f()));
  }), {
              v: 1,
              name: undefined,
              ast: {"params":[{"type":"i","name":"point"},{"type":"i","name":"dd"},{"type":"i","name":"halfHeight"}],"body":[0,[[13,"w",[6,[7,"d","vec2f"],["dd",[1,[6,[7,"std","abs"],[[7,"point","y"]]],"-","halfHeight"]]]],[10,[1,[6,[7,"std","min"],[[6,[7,"std","max"],[[7,"w","x"],[7,"w","y"]]],[5,"0"]]],"+",[6,[7,"std","length"],[[6,[7,"std","max"],["w",[6,[7,"d","vec2f"],[]]]]]]]]]],"externalNames":["d","std"]},
              externals: () => ({d, std}),
            }) && $.f)({}))
), "opExtrudeY"));

export { opExtrudeX, opExtrudeY, opExtrudeZ, opSmoothDifference, opSmoothUnion, opUnion, sdBezier, sdBezierApprox, sdBox2d, sdBox3d, sdBoxFrame3d, sdCapsule, sdDisk, sdLine, sdLine3d, sdPie, sdPlane, sdRoundedBox2d, sdRoundedBox3d, sdSphere };
